From ec65f6074830d31c60169530bb2f9595863b5ef7 Mon Sep 17 00:00:00 2001
From: Zhantao Tang <zhantao.tang@windriver.com>
Date: Thu, 12 May 2022 23:01:06 +0800
Subject: [PATCH] sw: fix dma share memory init failed issue

This patch is to fix the pfe dma share memory init failed issue, which is
caused by the kernel commit c5b3cecc1298 ("dts: fsl-s32g: pfe: Add
reserved-memory node for the routing table"), this commit adds one reserved
memory node for routing table of pfe, but did not add driver codes to
compatible with the dts update.

To fix the failure, there added a function to check the routing table
reserved memory node status, then the idx will be correct to the right
value and the init will pass.

Upstream-Status: Pending

Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>

---
 sw/oal/src/oal_mm_linux.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/sw/oal/src/oal_mm_linux.c b/sw/oal/src/oal_mm_linux.c
index 3bf411b..28f5339 100644
--- a/sw/oal/src/oal_mm_linux.c
+++ b/sw/oal/src/oal_mm_linux.c
@@ -553,6 +553,28 @@ out:
 }
 #endif
 
+#ifdef PFE_CFG_PFE_MASTER
+static int pfeng_reserved_rt_pool_region_check(struct device *dev, int idx)
+{
+	struct device_node *mem_node;
+
+	mem_node = of_parse_phandle(dev->of_node, "memory-region", idx);
+	if (!mem_node) {
+		dev_warn(dev, "No memory-region found at index %d\n", idx);
+		goto out;
+	}
+
+	if (!of_device_is_compatible(mem_node, "fsl,pfe-rt-pool")) {
+		dev_warn(dev, "fsl,pfe-rt-pool node missing\n");
+	}
+
+	of_node_put(mem_node);
+
+out:
+	return 0;
+}
+#endif
+
 errno_t oal_mm_init(const void *devh)
 {
 	struct device *dev = (struct device *)devh;
@@ -572,6 +594,10 @@ errno_t oal_mm_init(const void *devh)
 
 	if (rmem)
 		idx++;
+
+	pfeng_reserved_rt_pool_region_check(dev, idx);
+	idx++;
+	
 #endif /* PFE_CFG_PFE_MASTER */
 
 #if defined(PFE_CFG_PFE_MASTER) || defined(PFE_CFG_LINUX_RES_MEM_ENABLE)
-- 
2.25.1

